contracts/samples/TokenPaymaster.sol

```diff
@@ -1,113 +1,224 @@
+                    cachedPriceWithMarkup = clientSuppliedPrice;
+                }
+            }
+            uint256 tokenAmount = weiToToken(preChargeNative, cachedPriceWithMarkup);
+            SafeERC20.safeTransferFrom(token, userOp.sender, address(this), tokenAmount);
+            context = abi.encodePacked(tokenAmount, userOp.maxFeePerGas, userOp.maxPriorityFeePerGas, userOp.sender);
+            validationResult = 0;
+        }
     }
 
-    //Note: this method assumes a fixed ratio of token-to-eth. subclass should override to supply oracle
-    // or a setter.
-    function getTokenValueOfEth(uint256 valueEth) internal view virtual returns (uint256 valueToken) {
-        return valueEth / 100;
+    /// @notice Performs post-operation tasks, such as updating the token price and refunding excess tokens.
+    /// @dev This function is called after a user operation has been executed or reverted.
+    /// @param mode The post-operation mode (either successful or reverted).
+    /// @param context The context containing the token amount and user sender address.
+    /// @param actualGasCost The actual gas cost of the transaction.
+    function _postOp(PostOpMode mode, bytes calldata context, uint256 actualGasCost) internal override {
+        unchecked {
+            uint256 priceMarkup = tokenPaymasterConfig.priceMarkup;
+            uint256 preCharge = uint256(bytes32(context[0 : 32]));
+            uint256 maxFeePerGas = uint256(bytes32(context[32 : 64]));
+            uint256 maxPriorityFeePerGas = uint256(bytes32(context[64 : 96]));
+            uint256 gasPrice = getGasPrice(maxFeePerGas, maxPriorityFeePerGas);
+            address userOpSender = address(bytes20(context[96 : 116]));
+            if (mode == PostOpMode.postOpReverted) {
+                emit PostOpReverted(userOpSender, preCharge);
+                // Do nothing here to not revert the whole bundle and harm reputation
+                return;
+            }
+            uint256 _cachedPrice = updateCachedPrice(false);
+        // note: as price is in ether-per-token and we want more tokens increasing it means dividing it by markup
+            uint256 cachedPriceWithMarkup = _cachedPrice * PRICE_DENOMINATOR / priceMarkup;
+        // Refund tokens based on actual gas cost
+            uint256 actualChargeNative = actualGasCost + REFUND_POSTOP_COST * gasPrice;
+            uint256 actualTokenNeeded = weiToToken(actualChargeNative, cachedPriceWithMarkup);
+            if (preCharge > actualTokenNeeded) {
+                // If the initially provided token amount is greater than the actual amount needed, refund the difference
+                SafeERC20.safeTransfer(
+                    token,
+                    userOpSender,
+                    preCharge - actualTokenNeeded
+                );
+            } else if (preCharge < actualTokenNeeded) {
+                // Attempt to cover Paymaster's gas expenses by withdrawing the 'overdraft' from the client
```

Comment:
Not sure about this part. Won't this make `clientSuppliedPrice` invalid? `clientSuppliedPrice` should be the guard for how much token user is willing to pay but this part will end up forcing user to pay the full slippage

Reply:
Well, yes, with this implementation the user-supplied price can only increase the pre-charge for the user, but does not set an upper limit of the token price, so once the users approve this paymaster they basically agree to pay for gas at any current market price. 

@drortirosh what do you think about this feature?

CC: @huaweigu